oak_restricted_kernel/src/virtio_console.rs

```diff
@@ -32,69 +32,90 @@ use core::{
+unsafe impl Hal for OakHal {
+    fn dma_alloc(
+        pages: usize,
+        _direction: BufferDirection,
+    ) -> (virtio_drivers::PhysAddr, NonNull<u8>) {
+        let vaddr = GUEST_HOST_HEAP
+            .get()
+            .unwrap()
+            .allocate(Layout::from_size_align(pages * PAGE_SIZE, PAGE_SIZE).unwrap())
+            .expect("Failed to allocate memory for virtio MMIO")
+            .cast::<u8>();
+        let phys_addr = PAGE_TABLES
             .get()
             .unwrap()
             .lock()
-            .translate_virtual(vaddr)
+            .translate_virtual(VirtAddr::from_ptr(vaddr.as_ptr()))
             .unwrap()
-            .as_u64() as usize
+            .as_u64() as usize;
+        (phys_addr, vaddr)
     }
 
-    fn dma_dealloc(paddr: virtio_drivers::PhysAddr, pages: usize) -> i32 {
-        let vaddr = PAGE_TABLES
-            .get()
-            .unwrap()
-            .lock()
-            .translate_physical(PhysAddr::new(paddr as u64))
-            .unwrap();
-        unsafe {
-            GUEST_HOST_HEAP.get().unwrap().deallocate(
-                NonNull::new(vaddr.as_mut_ptr()).unwrap(),
-                Layout::from_size_align(pages * 0x1000, 0x1000).unwrap(),
-            );
-        }
+    unsafe fn dma_dealloc(
+        paddr: virtio_drivers::PhysAddr,
+        vaddr: NonNull<u8>,
+        pages: usize,
+    ) -> i32 {
+        let vaddr_check = Self::mmio_phys_to_virt(paddr, 0);
+        assert_eq!(
+            vaddr_check, vaddr,
+            "dma buffer physical and virtual addresses don't match",
+        );
+        GUEST_HOST_HEAP.get().unwrap().deallocate(
+            vaddr,
+            Layout::from_size_align(pages * 0x1000, 0x1000).unwrap(),
+        );
 
         0
     }
 
-    fn phys_to_virt(paddr: virtio_drivers::PhysAddr) -> virtio_drivers::VirtAddr {
-        PAGE_TABLES
-            .get()
-            .unwrap()
-            .lock()
-            .translate_physical(PhysAddr::new(paddr as u64))
-            .unwrap()
-            .as_u64() as usize
+    unsafe fn mmio_phys_to_virt(paddr: virtio_drivers::PhysAddr, _size: usize) -> NonNull<u8> {
+        NonNull::new(
+            PAGE_TABLES
+                .get()
+                .unwrap()
+                .lock()
+                .translate_physical(PhysAddr::new(paddr as u64))
+                .unwrap()
+                .as_mut_ptr(),
+        )
+        .unwrap()
     }
 
-    fn virt_to_phys(vaddr: virtio_drivers::VirtAddr) -> virtio_drivers::PhysAddr {
+    unsafe fn share(
+        buffer: NonNull<[u8]>,
+        _direction: BufferDirection,
+    ) -> virtio_drivers::PhysAddr {
+        // No additional work needed for sharing as the buffer was allocated from the guest-host
+        // allocator.
         PAGE_TABLES
             .get()
             .unwrap()
             .lock()
-            .translate_virtual(VirtAddr::new(vaddr as u64))
+            .translate_virtual(VirtAddr::from_ptr(buffer.cast::<u8>().as_ptr()))
```

Comment:
That cast seems clunky. Is it really required, as we only care about the fact that it's a pointer?

Reply:
`VirtAddr::from_ptr` has a trait bound that requires that the pointer is of a type with a size that is known at compile time, so the pointer to a slice does not work. The other option is `VirtAddr::new(buffer.as_ptr() as usize as u64)`, but that is also clunky.